packages/angular/ssr/src/app-engine.ts

```diff
@@ -0,0 +1,156 @@
+
+  /**
+   * Map of entry points from the `AngularAppEngine` manifest.
+   * Maps locale strings to their respective entry point paths.
+   */
+  private readonly entryPoints = new Map(Object.entries(this.manifest.entryPoints));
+
+  /**
+   * Renders an HTTP request using the appropriate Angular server application and returns a response.
+   *
+   * This method determines the entry point for the Angular server application based on the request URL,
+   * and caches the server application instances for reuse. If the application is in development mode,
+   * the cache is bypassed and a new instance is created for each request.
+   *
+   * If the request URL appears to be for a file (excluding `/index.html`), the method returns `null`.
+   * A request to `https://www.example.com/page/index.html` will render the Angular route
+   * corresponding to `https://www.example.com/page`.
+   *
+   * @param request - The incoming HTTP request object to be rendered.
+   * @param requestContext - Optional additional context for the request, such as metadata.
+   * @returns A promise that resolves to a Response object or null if the request is file-like.
+   */
+  async render(request: Request, requestContext?: unknown): Promise<Response | null> {
+    // Skip if the request looks like a file but not `/index.html`.
+    const url = new URL(request.url);
+    const { pathname } = url;
+    if (isFileLike(pathname) && !pathname.endsWith('/index.html')) {
+      return null;
+    }
+
+    const [locale, loadModule] = this.getEntryPointFromUrl(url);
+    let serverApp = this.appsCache.get(locale);
+    if (!serverApp) {
+      const { AngularServerApp } = await loadModule();
+      serverApp = new AngularServerApp({
+        isDevMode: AngularAppEngine.isDevMode,
+        hooks: this.hooks,
+      });
+
+      if (!AngularAppEngine.isDevMode) {
+        this.appsCache.set(locale, serverApp);
+      }
+    }
+
+    return serverApp.render(request, requestContext);
+  }
+
+  /**
+   * Retrieves the Angular server application's entry point path and locale based on the requested URL.
+   *
+   * @param url - The URL from which to derive the locale and determine the entry point.
+   * @returns An array where:
+   * - The first element is the locale derived from the URL.
+   * - The second element is a function that returns a promise resolving to an object containing the `AngularServerApp` type.
+   * @throws Error If the entry point path cannot be determined for the given URL.
+   */
+  private getEntryPointFromUrl(url: URL): [
+    locale: string,
+    loadModule: () => Promise<{
+      AngularServerApp: typeof AngularServerApp;
+    }>,
+  ] {
+    // Find bundle for locale
+    const entryPoints = this.entryPoints;
+    if (entryPoints.size === 1) {
+      return entryPoints.entries().next().value;
+    }
+
+    const potentialLocale = getPotentialLocaleIdFromUrl(url, this.manifest.basePath);
+    const entryPoint = entryPoints.get(potentialLocale);
```

Comment:
Question: What happens if I request `https://example.test/base/` directly (no locale)? If I understand this correctly it looks like it would look up an empty string entry point. Would that resolve to the source locale for the site?

Reply:
Managing these URLs will require a dedicated middleware because Angular localization includes the locale ID in the URL by default. Consequently, attempting to access a URL without a locale will result in a 404 error. This behavior mirrors that of client-side rendered applications with localization.

In the future, improvements could include detecting the locale from headers, allowing user-defined redirections, or using different domains for various language variants.

Reply:
> Consequently, attempting to access a URL without a locale will result in a 404 error.

That seems reasonable when there is a base URL which would require the user to follow some kind of redirect to reach in the first place (`https://example.test/base`), but that feels kind of bad if your app is hosted directly at the root path (`https://example.test/`). Would that mean that `https://example.test/` is a 404 by default and `https://example.test/en` would be required? I can't imagine any user who would want that behavior.

> This behavior mirrors that of client-side rendered applications with localization.

I'm all for aligning with existing behavior as much as possible. If this is an existing problem, then maybe it's something we don't worry about for now and look into changing it later?

Reply:
I agree that we should improve how we handle i18n redirects, domains, and paths. However, as mentioned, this is an existing problem that requires a separate, focused solution.